### PR TITLE
Problem: native action don't support events

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- manage the events and reverts

Closes: #XXX

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/statedb/statedb.go [hunk 1]
```diff
@@ -68,11 +68,14 @@ type StateDB struct {
...
 	accessList *accessList
+
+	// events emitted by native action
+	nativeEvents sdk.Events
 }
 
 // New creates a new state from a given trie.
 func New(ctx sdk.Context, keeper Keeper, txConfig TxConfig) *StateDB {
-	cacheCtx := ctx.WithMultiStore(cachemulti.NewStore(ctx.MultiStore(), keeper.StoreKeys()))
+	cacheCtx := ctx.WithMultiStore(cachemulti.NewStore(ctx.MultiStore(), keeper.StoreKeys())).WithEventManager(sdk.NewEventManager())
```
Comment (yihuang) [inline x/evm/statedb/statedb.go]:
```suggestion
	cacheCtx := ctx.WithMultiStore(cachemulti.NewStore(ctx.MultiStore(), keeper.StoreKeys()))
```

File: x/evm/statedb/statedb.go [hunk 2]
```diff
@@ -486,7 +497,7 @@ func (s *StateDB) Commit() error {
 	// commit the native cache store first,
 	// the states managed by precompiles and the other part of StateDB must not overlap.
 	s.CacheMultiStore().Write()
-        s.ctx.EventManager().EmitEvents(s.cacheCtx.EventManager().Events())
+	s.ctx.EventManager().EmitEvents(s.nativeEvents)
```
Comment (mmsqe) [inline x/evm/statedb/statedb.go]:
```suggestion
	if len(s.nativeEvents) > 0 {
		s.ctx.EventManager().EmitEvents(s.nativeEvents)
	}
```

File: x/evm/statedb/native.go [hunk 3]
```diff
@@ -17,4 +18,5 @@ func (native nativeChange) Dirtied() *common.Address {
 
 func (native nativeChange) Revert(s *StateDB) {
 	s.restoreNativeState(native.snapshot)
+	s.nativeEvents = s.nativeEvents[:len(s.nativeEvents)-native.events]
```
Comment (mmsqe) [inline x/evm/statedb/native.go:21]: if not assume out of bound
```suggestion
	if len(s.nativeEvents)-native.events >= 0 {
		s.nativeEvents = s.nativeEvents[:len(s.nativeEvents)-native.events]
	}
```

Comment (yihuang) [reply x/evm/statedb/native.go:21]: if `native.events > len(s.nativeEvents)`, there's something wrong.

Comment (yihuang) [reply x/evm/statedb/native.go:21]: if `native.events > len(s.nativeEvents)`, there's something serious wrong.

**Review by mmsqe — APPROVED** [review]

**Pull request merged**